acmeCA-2.0: Fix tests requiring Boulder container to avoid Docker failure

When running on builds, our Boulder using ACME fats intermittently hit 

`Caused by: com.github.dockerjava.api.exception.DockerException: {"message":"Pool overlaps with other one on this address space"}`

We need to work around this until we can write the Boulder container infrastructure.

The first plan is the split off tests from AcmeCaRestHandlerTest that actually require Boulder (like the Revoke tests) and leave everything else using Pebble. And then try add in some kind of sleep and retry to see if we can avoid collisions.

For #9017

--- a/dev/com.ibm.ws.security.acme/src/com/ibm/ws/security/acme/internal/AcmeCertCheckerTask.java
+++ b/dev/com.ibm.ws.security.acme/src/com/ibm/ws/security/acme/internal/AcmeCertCheckerTask.java
@@ -19,6 +19,8 @@ import java.util.concurrent.TimeUnit;
 
 import com.ibm.websphere.ras.Tr;
 import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.ffdc.annotation.FFDCIgnore;
+import com.ibm.wsspi.kernel.service.utils.FrameworkState;
 
 /**
  * The AcmeCertCheckerTask runs in the background and periodically checks if the
@@ -103,10 +105,18 @@ public class AcmeCertCheckerTask implements Runnable {
 	 * it continues to run, but is rescheduled on the error schedule.
 	 */
 	@Override
+	@FFDCIgnore(Throwable.class)
 	public void run() {
 
 		boolean isExpired = false, isRevoked = false;
 		List<X509Certificate> currentCert = null;
+		
+		if (FrameworkState.isStopping()) {
+			if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+				Tr.debug(tc, "Server is marked as stopping, cert checker returning.");
+			}
+			return;
+		}
 
 		acmeProviderImpl.acquireWriteLock();
 		try {
@@ -155,6 +165,12 @@ public class AcmeCertCheckerTask implements Runnable {
 			}
 
 		} catch (Throwable t) {
+			if (FrameworkState.isStopping()) {
+				if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+					Tr.debug(tc, "Caught an exception, but server is marked as stopping, cert checker returning.");
+				}
+				return;
+			}
 			try {
 				if (tc.isDebugEnabled()) {
 					Tr.debug(tc, "Requested a new certificate, but request failed.", t);

--- a/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/docker/CAContainer.java
+++ b/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/docker/CAContainer.java
@@ -52,6 +52,16 @@ public abstract class CAContainer extends GenericContainer<CAContainer> {
 	 * reachable.
 	 */
 	private final int dnsManagementPort;
+	
+	/**
+	 * Number of attemps to start containers again after catch an exception
+	 */
+	protected final static int NUM_RESTART_ATTEMPTS_ON_EXCEPTION = 3;
+	
+	/**
+	 * Value to put into withStartupAttempts on container config
+	 */
+	protected final static int WITH_STARTUP_ATTEMPTS = 20;
 
 	/**
 	 * Instantiate a new {@link CAContainer} instance.

--- a/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/docker/boulder/BoulderContainer.java
+++ b/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/docker/boulder/BoulderContainer.java
@@ -20,11 +20,11 @@ import org.testcontainers.Testcontainers;
 import org.testcontainers.containers.GenericContainer;
 import org.testcontainers.containers.Network;
 
+import com.github.dockerjava.api.exception.DockerException;
 import com.github.dockerjava.api.model.ContainerNetwork;
 import com.github.dockerjava.api.model.ContainerNetwork.Ipam;
 import com.ibm.websphere.simplicity.log.Log;
 import com.ibm.ws.security.acme.docker.CAContainer;
-import com.ibm.ws.security.acme.docker.pebble.PebbleContainer;
 
 /**
  * Testcontainer implementation for the letsencrypt/boulder container.
@@ -76,7 +76,7 @@ public class BoulderContainer extends CAContainer {
 	/**
 	 * Container that runs MariaDB
 	 */
-	public final GenericContainer<?> bmysql = new GenericContainer<>("mariadb:10.3").withNetwork(bluenet)
+	public final GenericContainer<?> bmysql = new GenericContainer<>("mariadb:10.3").withNetwork(bluenet).withNetworkMode("host")
 			.withExposedPorts(3306).withNetworkAliases("boulder-mysql").withEnv("MYSQL_ALLOW_EMPTY_PASSWORD", "yes")
 			.withCommand(
 					"mysqld --bind-address=0.0.0.0 --slow-query-log --log-output=TABLE --log-queries-not-using-indexes=ON")
@@ -117,9 +117,9 @@ public class BoulderContainer extends CAContainer {
 			throw new IllegalStateException("Failed to set default mock DNS A and AAAA record IP addresses.", e);
 		}
 
-		Log.info(PebbleContainer.class, "BoulderContainer", "ContainerIpAddress: " + getContainerIpAddress());
-		Log.info(PebbleContainer.class, "BoulderContainer", "DockerImageName:    " + getDockerImageName());
-		Log.info(PebbleContainer.class, "BoulderContainer", "ContainerInfo:      " + getContainerInfo());
+		Log.info(BoulderContainer.class, "BoulderContainer", "ContainerIpAddress: " + getContainerIpAddress());
+		Log.info(BoulderContainer.class, "BoulderContainer", "DockerImageName:    " + getDockerImageName());
+		Log.info(BoulderContainer.class, "BoulderContainer", "ContainerInfo:      " + getContainerInfo());
 	}
 
 	@Override
@@ -162,9 +162,106 @@ public class BoulderContainer extends CAContainer {
 		 * We need to start up the containers in an orderly fashion so that we
 		 * can pass the IP address of the DNS server to the Boulder server.
 		 */
-		bmysql.start();
-		bhsm.start();
-		super.start();
+		bmysql.withStartupAttempts(WITH_STARTUP_ATTEMPTS);
+		
+		for (int i = 1; i < NUM_RESTART_ATTEMPTS_ON_EXCEPTION + 1; i ++) {
+			try {
+				bmysql.start();
+				break;
+			} catch (Throwable t) {
+				Log.info(BoulderContainer.class, "start", "Failed to start bmysql, try again. " + t);
+				bmysql.stop();
+				Throwable cause = t.getCause();
+				while (cause != null) {
+					if (t instanceof DockerException) {
+						Log.info(BoulderContainer.class, "start", "Hit a Docker exception, trying a long sleep and retry");
+						try {
+							Thread.sleep(120000);
+						} catch (InterruptedException e) {
+						}
+						break;
+					}
+					cause = cause.getCause();
+				}
+
+				if (cause == null) { // do a short sleep and retry
+					try {
+						Thread.sleep(1000);
+					} catch (InterruptedException e) {
+					}
+				}
+			}
+		}
+		if (!bmysql.isRunning()) {
+			bmysql.start();
+		}
+		
+		bhsm.withStartupAttempts(WITH_STARTUP_ATTEMPTS);
+		for (int i = 1; i < NUM_RESTART_ATTEMPTS_ON_EXCEPTION + 1; i ++) {
+			try {
+				bhsm.start();
+				break;
+			} catch (Throwable t) {
+				Log.info(BoulderContainer.class, "start", "Failed to start bhsm, try again. " + t);
+				bhsm.stop();
+				Throwable cause = t.getCause();
+				while (cause != null) {
+					if (t instanceof DockerException) {
+						Log.info(BoulderContainer.class, "start", "Hit a Docker exception, trying a long sleep and retry");
+						try {
+							Thread.sleep(120000);
+						} catch (InterruptedException e) {
+						}
+						break;
+					}
+					cause = cause.getCause();
+				}
+
+				if (cause == null) { // do a short sleep and retry
+					try {
+						Thread.sleep(1000);
+					} catch (InterruptedException e) {
+					}
+				}
+			}
+		}
+		if (!bhsm.isRunning()) {
+			bhsm.start();
+		}
+		
+		super.withStartupAttempts(WITH_STARTUP_ATTEMPTS);
+		for (int i = 1; i < NUM_RESTART_ATTEMPTS_ON_EXCEPTION + 1; i ++) {
+			try {
+				super.start();
+				break;
+			} catch (Throwable t) {
+				Log.info(BoulderContainer.class, "start", "Failed to start boulder, try again. " + t);
+				super.stop();
+				Throwable cause = t.getCause();
+				while (cause != null) {
+					if (t instanceof DockerException) {
+						Log.info(BoulderContainer.class, "start", "Hit a Docker exception, trying a long sleep and retry");
+						try {
+							Thread.sleep(120000);
+						} catch (InterruptedException e) {
+						}
+						break;
+					}
+					cause = cause.getCause();
+				}
+
+				if (cause == null) { // do a short sleep and retry
+					try {
+						Thread.sleep(1000);
+					} catch (InterruptedException e) {
+					}
+				}
+			}
+		}
+		if (!super.isRunning()) {
+			super.start();
+		}
+
 	}
 
 	@Override

--- a/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/docker/pebble/PebbleContainer.java
+++ b/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/docker/pebble/PebbleContainer.java
@@ -13,6 +13,7 @@ package com.ibm.ws.security.acme.docker.pebble;
 
 import static junit.framework.Assert.fail;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
 
 import java.io.File;
 import java.io.IOException;
@@ -25,6 +26,7 @@ import org.testcontainers.containers.Network;
 import org.testcontainers.containers.output.OutputFrame;
 import org.testcontainers.images.builder.ImageFromDockerfile;
 
+import com.github.dockerjava.api.exception.DockerException;
 import com.github.dockerjava.api.model.ContainerNetwork;
 import com.ibm.websphere.simplicity.log.Log;
 import com.ibm.ws.security.acme.docker.CAContainer;
@@ -80,8 +82,49 @@ public class PebbleContainer extends CAContainer {
 						.copy("pebble-config.json", "/test/config/pebble-config.json").build())
 				.withFileFromFile("pebble-config.json", new File("lib/LibertyFATTestFiles/pebble-config.json")), 5002,
 				14000, 15000);
+		challtestsrv.withStartupAttempts(20);
+		challtestsrv.withStartupTimeout(Duration.ofSeconds(60));
 
-		challtestsrv.start();
+		for (int i = 1; i < NUM_RESTART_ATTEMPTS_ON_EXCEPTION + 1; i ++) {
+			try {
+				challtestsrv.start();
+				Log.info(PebbleContainer.class, "PebbleContainer", "challtestsrv started");
+				break;
+			} catch (Throwable t) {
+				Log.info(PebbleContainer.class, "PebbleContainer", "Failed to start challtestsrv, try again. " + t);
+				challtestsrv.stop();
+				Throwable cause = t.getCause();
+				while (cause != null) {
+					if (t instanceof DockerException) {
+						Log.info(PebbleContainer.class, "PebbleContainer", "Hit a Docker exception, trying a long sleep and retry");
+						try {
+							Thread.sleep(120000);
+						} catch (InterruptedException e) {
+						}
+						break;
+					}
+					cause = cause.getCause();
+				}
+
+				if (cause == null) { // do a short sleep and retry
+					try {
+						Thread.sleep(1000);
+					} catch (InterruptedException e) {
+					}
+				}
+			}
+		}
+		if (!challtestsrv.isRunning()) {
+			challtestsrv.start();
+			Log.info(PebbleContainer.class, "PebbleContainer", "challtestsrv started");
+			
+			/**
+			 * Intermittently getting a null calling getIntraContainerIP, determine what is null
+			 */
+			assertNotNull("challtestsrv.getContainerInfo()", challtestsrv.getContainerInfo());
+			assertNotNull("challtestsrv.getContainerInfo().getNetworkSettings()", challtestsrv.getContainerInfo().getNetworkSettings());
+			assertNotNull("challtestsrv.getContainerInfo().getNetworkSettings().getNetworks().entrySet()", challtestsrv.getContainerInfo().getNetworkSettings().getNetworks().entrySet());
+		}
 
 		String dnsServer = getIntraContainerIP() + ":" + DNS_PORT;
 
@@ -90,12 +133,43 @@ public class PebbleContainer extends CAContainer {
 		this.withExposedPorts(getDnsManagementPort(), getAcmeListenPort());
 		this.withNetwork(network);
 		this.withLogConsumer(PebbleContainer::log);
-		this.withStartupAttempts(3);
+		this.withStartupAttempts(20);
 		this.withStartupTimeout(Duration.ofSeconds(60));
 
 		Testcontainers.exposeHostPorts(5002);
 
-		start();
+		for (int i = 1; i < NUM_RESTART_ATTEMPTS_ON_EXCEPTION + 1; i ++) {
+			try {
+				start();
+				break;
+			} catch (Throwable t) {
+				Log.info(PebbleContainer.class, "PebbleContainer", "Failed to start pebble, try again. " + t);
+				super.stop();
+				Throwable cause = t.getCause();
+				while (cause != null) {
+					if (t instanceof DockerException) {
+						Log.info(PebbleContainer.class, "PebbleContainer", "Hit a Docker exception, trying a long sleep and retry");
+						try {
+							Thread.sleep(120000);
+						} catch (InterruptedException e) {
+						}
+						break;
+					}
+					cause = cause.getCause();
+				}
+
+				if (cause == null) { // do a short sleep and retry
+					try {
+						Thread.sleep(1000);
+					} catch (InterruptedException e) {
+					}
+				}
+			}
+		}
+		
+		if (!isRunning()) {
+			start();
+		}
 
 		try {
 			/*
@@ -115,6 +189,7 @@ public class PebbleContainer extends CAContainer {
 
 		Log.info(PebbleContainer.class, "PebbleContainer", "ContainerIpAddress: " + getContainerIpAddress());
 		Log.info(PebbleContainer.class, "PebbleContainer", "DockerImageName:    " + getDockerImageName());
+		assertNotNull("getContainerInfo()", getContainerInfo());
 		Log.info(PebbleContainer.class, "PebbleContainer", "ContainerInfo:      " + getContainerInfo());
 	}
 

--- a/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/fat/AcmeCaRestHandlerTest.java
+++ b/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/fat/AcmeCaRestHandlerTest.java
@@ -53,7 +53,6 @@ import org.shredzone.acme4j.util.KeyPairUtils;
 import com.ibm.websphere.simplicity.config.ServerConfiguration;
 import com.ibm.websphere.simplicity.log.Log;
 import com.ibm.ws.security.acme.docker.CAContainer;
-import com.ibm.ws.security.acme.docker.boulder.BoulderContainer;
 import com.ibm.ws.security.acme.docker.pebble.PebbleContainer;
 import com.ibm.ws.security.acme.internal.util.AcmeConstants;
 import com.ibm.ws.security.acme.internal.web.AcmeCaRestHandler;
@@ -89,7 +88,7 @@ public class AcmeCaRestHandlerTest {
 	private static final String UNAUTHORIZED_PASS = "unauthorizedpass";
 
 	@ClassRule
-	public static CAContainer boulder = new BoulderContainer();
+	public static CAContainer pebble = new PebbleContainer();
 
 	private static final String JSON_ACCOUNT_REGEN_KEYPAIR_VALID = "{\"" + AcmeCaRestHandler.OP_KEY + "\":\""
 			+ AcmeCaRestHandler.OP_RENEW_ACCT_KEY_PAIR + "\"}";
@@ -100,13 +99,6 @@ public class AcmeCaRestHandlerTest {
 	private static final String JSON_CERT_REGEN = "{\"" + AcmeCaRestHandler.OP_KEY + "\":\""
 			+ AcmeCaRestHandler.OP_RENEW_CERT + "\"}";
 	private static final String JSON_CERT_INVALID_OP = "{\"" + AcmeCaRestHandler.OP_KEY + "\":\"invalid\"}";
-
-	private static final String JSON_CERT_REVOKE_DEFAULT_REASON = "{\"" + AcmeCaRestHandler.OP_KEY + "\":\""
-			+ AcmeCaRestHandler.OP_REVOKE_CERT + "\"}";
-	private static final String JSON_CERT_REVOKE_VALID_REASON = "{\"" + AcmeCaRestHandler.OP_KEY + "\":\""
-			+ AcmeCaRestHandler.OP_REVOKE_CERT + "\",\"" + AcmeCaRestHandler.REASON_KEY + "\":\"superseded\"}";
-	private static final String JSON_CERT_REVOKE_INVALID_REASON = "{\"" + AcmeCaRestHandler.OP_KEY + "\":\""
-			+ AcmeCaRestHandler.OP_REVOKE_CERT + "\",\"" + AcmeCaRestHandler.REASON_KEY + "\":\"invalid\"}";
 
 	private static final String CONTENT_TYPE_HTML = "text/html";
 	private static final String CONTENT_TYPE_JSON = "application/json";
@@ -125,12 +117,12 @@ public class AcmeCaRestHandlerTest {
 		/*
 		 * Make sure the HTTP port is open.
 		 */
-		AcmeFatUtils.checkPortOpen(boulder.getHttpPort(), 60000);
+		AcmeFatUtils.checkPortOpen(pebble.getHttpPort(), 60000);
 
 		/*
 		 * Configure the acmeCA-2.0 feature.
 		 */
-		AcmeFatUtils.configureAcmeCA(server, boulder, ORIGINAL_CONFIG, false, true, DOMAINS);
+		AcmeFatUtils.configureAcmeCA(server, pebble, ORIGINAL_CONFIG, false, true, DOMAINS);
 
 		server.startServer();
 		AcmeFatUtils.waitForAcmeToCreateCertificate(server);
@@ -301,39 +293,6 @@ public class AcmeCaRestHandlerTest {
 		String serial1 = getLeafSerialFromHtml(html1);
 		String serial2 = getLeafSerialFromHtml(html2);
 		assertThat("Certificates should have been different.", serial2, not(equalTo(serial1)));
-	}
-
-	@Test
-	public void certificate_endpoint_post_revoke_certificate() throws Exception {
-		/*
-		 * Revoke with invalid reason.
-		 */
-		String jsonResponse = performPost(CERTIFICATE_ENDPOINT, 400, CONTENT_TYPE_JSON, ADMIN_USER, ADMIN_PASS,
-				CONTENT_TYPE_JSON, JSON_CERT_REVOKE_INVALID_REASON);
-		assertJsonResponse(jsonResponse, 400);
-
-		/*
-		 * Revoke with unspecified / default reason.
-		 */
-		jsonResponse = performPost(CERTIFICATE_ENDPOINT, 200, CONTENT_TYPE_JSON, ADMIN_USER, ADMIN_PASS,
-				CONTENT_TYPE_JSON, JSON_CERT_REVOKE_DEFAULT_REASON);
-		assertJsonResponse(jsonResponse, 200);
-
-		/*
-		 * Request a new certificate.
-		 */
-		jsonResponse = performPost(CERTIFICATE_ENDPOINT, 200, CONTENT_TYPE_JSON, ADMIN_USER, ADMIN_PASS,
-				CONTENT_TYPE_JSON, JSON_CERT_REGEN);
-		assertJsonResponse(jsonResponse, 200);
-		AcmeFatUtils.waitForAcmeToCreateCertificate(server);
-		AcmeFatUtils.waitForSslEndpoint(server);
-
-		/*
-		 * Revoke with a valid reason.
-		 */
-		jsonResponse = performPost(CERTIFICATE_ENDPOINT, 200, CONTENT_TYPE_JSON, ADMIN_USER, ADMIN_PASS,
-				CONTENT_TYPE_JSON, JSON_CERT_REVOKE_VALID_REASON);
-		assertJsonResponse(jsonResponse, 200);
 	}
 
 	@Test
@@ -848,12 +807,12 @@ public class AcmeCaRestHandlerTest {
 			/*
 			 * First renew request should update.
 			 */
-			startingCertificateChain = AcmeFatUtils.assertAndGetServerCertificate(server, boulder);
+			startingCertificateChain = AcmeFatUtils.assertAndGetServerCertificate(server, pebble);
 
 			String jsonResponse = performPost(AcmeCaRestHandlerTest.CERTIFICATE_ENDPOINT, 200, AcmeCaRestHandlerTest.CONTENT_TYPE_JSON, AcmeCaRestHandlerTest.ADMIN_USER, AcmeCaRestHandlerTest.ADMIN_PASS,
 					AcmeCaRestHandlerTest.CONTENT_TYPE_JSON, AcmeCaRestHandlerTest.JSON_CERT_REGEN);
 			assertJsonResponse(jsonResponse, 200);
-			AcmeFatUtils.waitForNewCert(server, boulder, startingCertificateChain);
+			AcmeFatUtils.waitForNewCert(server, pebble, startingCertificateChain);
 			
 			/*
 			 * Do back to back renew requests, we should be blocked from renewing
@@ -863,7 +822,7 @@ public class AcmeCaRestHandlerTest {
 				
 				Log.info(this.getClass(), testName.getMethodName(), "Renew round " + i);
 				
-				startingCertificateChain = AcmeFatUtils.assertAndGetServerCertificate(server, boulder);
+				startingCertificateChain = AcmeFatUtils.assertAndGetServerCertificate(server, pebble);
 
 				jsonResponse = performPost(AcmeCaRestHandlerTest.CERTIFICATE_ENDPOINT, 429, AcmeCaRestHandlerTest.CONTENT_TYPE_JSON, AcmeCaRestHandlerTest.ADMIN_USER, AcmeCaRestHandlerTest.ADMIN_PASS,
 						AcmeCaRestHandlerTest.CONTENT_TYPE_JSON, AcmeCaRestHandlerTest.JSON_CERT_REGEN);
@@ -873,7 +832,7 @@ public class AcmeCaRestHandlerTest {
 				
 				Log.info(this.getClass(), methodName, "Response received: " + jsonResponse);
 				
-				endingCertificateChain = AcmeFatUtils.assertAndGetServerCertificate(server, boulder);
+				endingCertificateChain = AcmeFatUtils.assertAndGetServerCertificate(server, pebble);
 
 				assertEquals("The certificate should not renew after REST request.",
 						((X509Certificate) startingCertificateChain[0]).getSerialNumber(),
@@ -885,13 +844,13 @@ public class AcmeCaRestHandlerTest {
 			 */
 			Thread.sleep(AcmeConstants.RENEW_CERT_MIN + 2000);
 			
-			startingCertificateChain = AcmeFatUtils.assertAndGetServerCertificate(server, boulder);
+			startingCertificateChain = AcmeFatUtils.assertAndGetServerCertificate(server, pebble);
 
 			jsonResponse = performPost(AcmeCaRestHandlerTest.CERTIFICATE_ENDPOINT, 200, AcmeCaRestHandlerTest.CONTENT_TYPE_JSON, AcmeCaRestHandlerTest.ADMIN_USER, AcmeCaRestHandlerTest.ADMIN_PASS,
 					AcmeCaRestHandlerTest.CONTENT_TYPE_JSON, AcmeCaRestHandlerTest.JSON_CERT_REGEN);
 			assertJsonResponse(jsonResponse, 200);
 			
-			AcmeFatUtils.waitForNewCert(server, boulder, startingCertificateChain);
+			AcmeFatUtils.waitForNewCert(server, pebble, startingCertificateChain);
 
 		} finally {
 			/*

--- a/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/fat/AcmeRevocationTest.java
+++ b/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/fat/AcmeRevocationTest.java
@@ -13,11 +13,14 @@ package com.ibm.ws.security.acme.fat;
 
 import static junit.framework.Assert.assertEquals;
 import static junit.framework.Assert.assertNotNull;
+import static org.hamcrest.CoreMatchers.containsString;
 import static org.hamcrest.CoreMatchers.equalTo;
 import static org.hamcrest.CoreMatchers.not;
 import static org.junit.Assert.assertNotSame;
 import static org.junit.Assert.assertThat;
 
+import java.io.File;
+import java.io.FileFilter;
 import java.math.BigInteger;
 import java.security.cert.Certificate;
 import java.security.cert.X509Certificate;
@@ -84,6 +87,21 @@ public class AcmeRevocationTest {
 	private static final String DOMAIN = "domain1.com"; // Set to domain1.com as it has a high rate limit on certificate requests
 	
 	public static final long SCHEDULE_TIME = AcmeConstants.RENEW_CERT_MIN + 2000;
+	
+	private static final String CERTIFICATE_ENDPOINT = "/ibm/api" + AcmeCaRestHandler.PATH_CERTIFICATE;
+	
+	private static final String JSON_CERT_REGEN = "{\"" + AcmeCaRestHandler.OP_KEY + "\":\""
+			+ AcmeCaRestHandler.OP_RENEW_CERT + "\"}";
+	private static final String JSON_CERT_INVALID_OP = "{\"" + AcmeCaRestHandler.OP_KEY + "\":\"invalid\"}";
+
+	private static final String JSON_CERT_REVOKE_DEFAULT_REASON = "{\"" + AcmeCaRestHandler.OP_KEY + "\":\""
+			+ AcmeCaRestHandler.OP_REVOKE_CERT + "\"}";
+	private static final String JSON_CERT_REVOKE_VALID_REASON = "{\"" + AcmeCaRestHandler.OP_KEY + "\":\""
+			+ AcmeCaRestHandler.OP_REVOKE_CERT + "\",\"" + AcmeCaRestHandler.REASON_KEY + "\":\"superseded\"}";
+	private static final String JSON_CERT_REVOKE_INVALID_REASON = "{\"" + AcmeCaRestHandler.OP_KEY + "\":\""
+			+ AcmeCaRestHandler.OP_REVOKE_CERT + "\",\"" + AcmeCaRestHandler.REASON_KEY + "\":\"invalid\"}";
+
+	private static final String CONTENT_TYPE_JSON = "application/json";
 
 	@Rule
 	public TestName testName = new TestName();
@@ -416,6 +434,156 @@ public class AcmeRevocationTest {
 			 * Stop the server.
 			 */
 			server.stopServer();
+		}
+	}
+	
+	/**
+	 * Test the REST revoke endpoint, provide valid and invalid reasons for the revoke
+	 * @throws Exception
+	 */
+	@Test
+	public void certificate_endpoint_post_revoke_certificate() throws Exception {
+		/*
+		 * Configure the acmeCA-2.0 feature.
+		 */
+		ServerConfiguration configuration = originalServerConfig.clone();
+		configureAcmeRevocation(configuration, true);
+		AcmeFatUtils.configureAcmeCA(server, boulder, configuration, false, true, DOMAIN);
+
+		try {
+
+			/*
+			 * Startup the server.
+			 */
+			Log.info(this.getClass(), testName.getMethodName(), "Rest revokation test.");
+			server.startServer();
+			AcmeFatUtils.waitForSslToCreateKeystore(server);
+			AcmeFatUtils.waitForSslEndpoint(server);
+
+
+			/*
+			 * Revoke with invalid reason.
+			 */
+			String jsonResponse = performPost(CERTIFICATE_ENDPOINT, 400, CONTENT_TYPE_JSON, ADMIN_USER, ADMIN_PASS,
+					CONTENT_TYPE_JSON, JSON_CERT_REVOKE_INVALID_REASON);
+			assertJsonResponse(jsonResponse, 400);
+
+			/*
+			 * Revoke with unspecified / default reason.
+			 */
+			jsonResponse = performPost(CERTIFICATE_ENDPOINT, 200, CONTENT_TYPE_JSON, ADMIN_USER, ADMIN_PASS,
+					CONTENT_TYPE_JSON, JSON_CERT_REVOKE_DEFAULT_REASON);
+			assertJsonResponse(jsonResponse, 200);
+
+			/*
+			 * Request a new certificate.
+			 */
+			jsonResponse = performPost(CERTIFICATE_ENDPOINT, 200, CONTENT_TYPE_JSON, ADMIN_USER, ADMIN_PASS,
+					CONTENT_TYPE_JSON, JSON_CERT_REGEN);
+			assertJsonResponse(jsonResponse, 200);
+			AcmeFatUtils.waitForAcmeToCreateCertificate(server);
+			AcmeFatUtils.waitForSslEndpoint(server);
+
+			/*
+			 * Revoke with a valid reason.
+			 */
+			jsonResponse = performPost(CERTIFICATE_ENDPOINT, 200, CONTENT_TYPE_JSON, ADMIN_USER, ADMIN_PASS,
+					CONTENT_TYPE_JSON, JSON_CERT_REVOKE_VALID_REASON);
+			assertJsonResponse(jsonResponse, 200);
+
+		} finally {
+			server.stopServer();
+		}
+	}
+
+	/**
+	 * Make a POST call to the REST endpoint.
+	 * 
+	 * @param endpoint            The endpoint to call.
+	 * @param expectedStatus      The expected HTTP return code.
+	 * @param expectedContentType The expected response content type.
+	 * @param user                The user to make the call with.
+	 * @param password            The password to make the call with.
+	 * @return the response in the form of a string.
+	 * @throws Exception If the call failed.
+	 */
+	private static String performPost(String endpoint, int expectedStatus, String expectedContentType, String user,
+			String password, String contentType, String content) throws Exception {
+		String methodName = "performPost()";
+
+		try (CloseableHttpClient httpclient = AcmeFatUtils.getInsecureHttpsClient()) {
+
+			/*
+			 * Create a POST request to the Liberty server.
+			 */
+			HttpPost httpPost = new HttpPost("https://localhost:" + server.getHttpDefaultSecurePort() + endpoint);
+			httpPost.setHeader("Authorization",
+					"Basic " + DatatypeConverter.printBase64Binary((user + ":" + password).getBytes()));
+			if (contentType != null && !contentType.isEmpty()) {
+				httpPost.setHeader("Content-Type", contentType);
+			}
+			if (content != null && !content.isEmpty()) {
+				httpPost.setEntity(new StringEntity(content));
+			}
+
+			/*
+			 * Send the POST request and process the response.
+			 */
+			try (final CloseableHttpResponse response = httpclient.execute(httpPost)) {
+				AcmeFatUtils.logHttpResponse(PebbleContainer.class, methodName, httpPost, response);
+
+				StatusLine statusLine = response.getStatusLine();
+				assertEquals("Unexpected status code response.", expectedStatus, statusLine.getStatusCode());
+
+				/*
+				 * Check content type header.
+				 */
+				if (expectedContentType != null) {
+					Header[] headers = response.getHeaders("content-type");
+					assertNotNull("Expected content type header.", headers);
+					assertEquals("Expected 1 content type header.", 1, headers.length);
+					assertEquals("Unexpected content type.", expectedContentType, headers[0].getValue());
+				}
+
+				String contentString = EntityUtils.toString(response.getEntity());
+				Log.info(AcmeRevocationTest.class, methodName, "HTTP post contents: \n" + contentString);
+
+				return contentString;
+			}
+		}
+	}
+
+	/**
+	 * Find all files in the directory whose file name matches the given pattern.
+	 * 
+	 * @param dir     The directory to search.
+	 * @param pattern The pattern to check file names against.
+	 * @return The array of files that match.
+	 */
+	public static File[] findFilesThatMatch(String dir, final String pattern) {
+		File searchDir = new File(dir);
+
+		return searchDir.listFiles(new FileFilter() {
+			public boolean accept(File pathname) {
+				return pathname.getName().matches(pattern);
+			}
+		});
+	}
+
+	/**
+	 * Assert that the expected JSON response is returned.
+	 * 
+	 * @param jsonResponse       The actual JSON response.
+	 * @param expectedStatusCode The expected status code.
+	 */
+	private static void assertJsonResponse(String jsonResponse, int expectedStatusCode) {
+		assertThat("Unexpected HTTP status code returned in JSON response.", jsonResponse,
+				containsString("\"httpCode\":" + expectedStatusCode));
+		if (expectedStatusCode == 200) {
+			assertThat("Expected no error message in JSON response.", jsonResponse,
+					not(containsString("\"message\":")));
+		} else {
+			assertThat("Expected error message in JSON response.", jsonResponse, containsString("\"message\":"));
 		}
 	}
 

--- a/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/fat/AcmeValidityAndRenewTest.java
+++ b/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/fat/AcmeValidityAndRenewTest.java
@@ -311,6 +311,7 @@ public class AcmeValidityAndRenewTest {
 		 * TEST 2: Start with a very short renew period, causing a renewal on startup.
 		 * 
 		 **********************************************************************/
+		long justShyOfValidityPeriod =0;
 		try {
 			Log.info(this.getClass(), testName.getMethodName(),
 					"TEST 2: Restart with renew time close to validity period.");
@@ -330,7 +331,7 @@ public class AcmeValidityAndRenewTest {
 			 * Set a renew time just shy of default Pebble validity period) so we will
 			 * request a new certificate on restart)
 			 */
-			long justShyOfValidityPeriod = (notAfter - notBefore) - timeBufferToExpire;
+			 justShyOfValidityPeriod = (notAfter - notBefore) - timeBufferToExpire;
 			Log.info(this.getClass(), testName.getMethodName(), "Time configured: " + justShyOfValidityPeriod);
 
 			/*
@@ -381,6 +382,7 @@ public class AcmeValidityAndRenewTest {
 			Log.info(this.getClass(), testName.getMethodName(),
 					"TEST 3: Restart with renew time close to validity period and short cert checker.");
 
+			configuration.getAcmeCA().setRenewBeforeExpiration((justShyOfValidityPeriod +5000) + "ms");
 			configuration.getAcmeCA().setCertCheckerSchedule((timeBufferToExpire + 1000) + "ms");
 			configuration.getAcmeCA().setCertCheckerSchedule(AcmeConstants.RENEW_CERT_MIN + "ms");
 
@@ -405,10 +407,22 @@ public class AcmeValidityAndRenewTest {
 			/*
 			 * The certificate checker should automatically renew the certificate.
 			 */
-			AcmeFatUtils.waitForNewCert(server, caContainer, startingCertificateChain, timeBufferToExpire * 2);
+			AcmeFatUtils.waitForNewCert(server, caContainer, startingCertificateChain, timeBufferToExpire * 4);
 
 			assertNotNull("Should log message that the certificate was renewed",
 					server.waitForStringInLogUsingMark("CWPKI2052I"));
+			
+			/**
+			 * Run "load" while the certificate checkers runs in the background and renews the cert
+			 */
+			long now = System.currentTimeMillis();
+			Log.info(this.getClass(), testName.getMethodName(), "Run https load to make sure certificate renew happens cleanly.");
+			while (System.currentTimeMillis() - now < timeBufferToExpire * 2) {
+				AcmeFatUtils.assertAndGetServerCertificate(server, caContainer);
+			}
+			
+			assertTrue("Should not find CWPKI0033E", server.findStringsInLogs("CWPKI0033E").isEmpty());
+			assertTrue("Should not find CWPKI0809W", server.findStringsInLogs("CWPKI0809W").isEmpty());
 
 		} finally {
 			Log.info(this.getClass(), testName.getMethodName(), "TEST 3: Shutdown.");
@@ -626,8 +640,11 @@ public class AcmeValidityAndRenewTest {
 			 */
 			caContainer.stopDNSServer();
 
+			/*
+			 * DNS info can be cached, give more time for cache to expire
+			 */
 			assertNotNull("Should log message that the certificate renew failed",
-					server.waitForStringInLogUsingMark("CWPKI2065W", timeBufferToExpire * 2));
+					server.waitForStringInLogUsingMark("CWPKI2065W", timeBufferToExpire * 4));
 
 			/*
 			 * Start the Challenge server and the cert checker should recover and fetch a
@@ -647,7 +664,7 @@ public class AcmeValidityAndRenewTest {
 			}
 
 			assertNotNull("Should log message that the certificate was renewed after restarting the challenge server",
-					server.waitForStringInLogUsingMark("CWPKI2007I", (AcmeConstants.RENEW_CERT_MIN * 2)));
+					server.waitForStringInLogUsingMark("CWPKI2007I", (AcmeConstants.RENEW_CERT_MIN * 6)));
 			AcmeFatUtils.waitForNewCert(server, caContainer, startingCertificateChain, timeBufferToExpire);
 
 		} finally {

--- a/dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/config/WSKeyStore.java
+++ b/dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/config/WSKeyStore.java
@@ -10,6 +10,8 @@
  *******************************************************************************/
 package com.ibm.ws.ssl.config;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
@@ -38,12 +40,15 @@ import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Properties;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 import com.ibm.websphere.crypto.InvalidPasswordDecodingException;
 import com.ibm.websphere.crypto.PasswordUtil;
 import com.ibm.websphere.ras.Tr;
 import com.ibm.websphere.ras.TraceComponent;
 import com.ibm.websphere.ras.annotation.Sensitive;
+import com.ibm.websphere.ras.annotation.Trivial;
 import com.ibm.websphere.ssl.Constants;
 import com.ibm.websphere.ssl.JSSEProvider;
 import com.ibm.websphere.ssl.SSLException;
@@ -117,6 +122,9 @@ public class WSKeyStore extends Properties {
     private static final String SUNPKCS11_PROVIDER_NAME = "SunPKCS11";
 
     private final Map<String, SerializableProtectedString> certAliasInfo = new HashMap<String, SerializableProtectedString>();
+
+    /** Read/Write lock to prevent multiple processes writing the keystore file at the same time, or reading while we are writing. Added for acmeCA feature. **/
+    private final ReadWriteLock rwKeyStoreLock = new ReentrantReadWriteLock();
 
     /**
      * Default constructor, will initialize default values.
@@ -745,7 +753,7 @@ public class WSKeyStore extends Properties {
      * @return KeyStore
      * @throws Exception
      */
-    public synchronized KeyStore do_getKeyStore(boolean reinitialize, boolean createIfNotPresent) throws Exception {
+    private synchronized KeyStore do_getKeyStore(boolean reinitialize, boolean createIfNotPresent) throws Exception {
         if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled())
             Tr.entry(tc, "do_getKeyStore", new Object[] { Boolean.valueOf(reinitialize), Boolean.valueOf(createIfNotPresent) });
 
@@ -1002,11 +1010,16 @@ public class WSKeyStore extends Properties {
      * @throws Exception
      */
     public KeyStore getKeyStore(boolean reinitialize, boolean createIfNotPresent) throws Exception {
-        if (myKeyStore == null || reinitialize) {
-            myKeyStore = do_getKeyStore(reinitialize, createIfNotPresent);
-        }
+        acquireReadLock();
+        try {
+            if (myKeyStore == null || reinitialize) {
+                myKeyStore = do_getKeyStore(reinitialize, createIfNotPresent);
+            }
 
-        return myKeyStore;
+            return myKeyStore;
+        } finally {
+            releaseReadLock();
+        }
     }
 
     /**
@@ -1017,6 +1030,8 @@ public class WSKeyStore extends Properties {
     public void store() throws Exception {
         if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled())
             Tr.entry(tc, "store");
+
+        acquireWriteLock();
 
         try {
             String name = getProperty(Constants.SSLPROP_KEY_STORE_NAME);
@@ -1067,6 +1082,8 @@ public class WSKeyStore extends Properties {
                 Tr.debug(tc, "Exception storing KeyStore; " + e);
             FFDCFilter.processException(e, getClass().getName(), "store", this);
             throw e;
+        } finally {
+            releaseWriteLock();
         }
 
         if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled())
@@ -1276,9 +1293,9 @@ public class WSKeyStore extends Properties {
      * @param alias
      * @param cert
      * @throws KeyStoreException
-     *             - if the store is read only or not found
+     *                               - if the store is read only or not found
      * @throws KeyException
-     *             - if an error happens updating the store with the cert
+     *                               - if an error happens updating the store with the cert
      */
     public void setCertificateEntry(String alias, Certificate cert) throws KeyStoreException, KeyException {
         if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
@@ -1362,9 +1379,9 @@ public class WSKeyStore extends Properties {
      * @param password
      * @param chain
      * @throws KeyStoreException
-     *             - if the store is read only or not found
+     *                               - if the store is read only or not found
      * @throws KeyException
-     *             - if an error happens updating the store with the cert
+     *                               - if an error happens updating the store with the cert
      */
     @Sensitive
     public void setKeyEntry(String alias, Key key, Certificate[] chain) throws KeyStoreException, KeyException {
@@ -1546,15 +1563,15 @@ public class WSKeyStore extends Properties {
     }
 
     /**
-     * Set a new certificate into the keystore and save the updated store.
+     * Set a new certificate into the keystore
      *
      * @param alias
      *
      * @param cert
      * @throws KeyStoreException
-     *             - if the store is read only or not found
+     *                               - if the store is read only or not found
      * @throws KeyException
-     *             - if an error happens updating the store with the cert
+     *                               - if an error happens updating the store with the cert
      */
     private void setCertificateEntryNoStore(String alias, Certificate cert) throws KeyStoreException, KeyException {
         if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
@@ -1582,7 +1599,12 @@ public class WSKeyStore extends Properties {
     protected void clearJavaKeyStore() {
         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
             Tr.debug(tc, "clearJavaKeyStore");
-        myKeyStore = null;
+        acquireWriteLock();
+        try {
+            myKeyStore = null;
+        } finally {
+            releaseWriteLock();
+        }
     }
 
     public static String getCannonicalPath(String location, Boolean fileBased) {
@@ -1628,4 +1650,62 @@ public class WSKeyStore extends Properties {
         return (ext);
     }
 
+    /**
+     * Acquire the writer lock. To be used to prevent concurrent keystore
+     * fetching and storing. Must be used with releaseWriteLock
+     */
+    @Trivial
+    void acquireWriteLock() {
+        rwKeyStoreLock.writeLock().lock();
+    }
+
+    /**
+     * Release the writer lock. To be used to prevent concurrent keystore
+     * fetching and storing. Must be used with acquireWriteLock
+     */
+    @Trivial
+    void releaseWriteLock() {
+        rwKeyStoreLock.writeLock().unlock();
+    }
+
+    /**
+     * Acquire the reader lock. To be used to prevent concurrent keystore
+     * fetching and storing. Must be used with releaseReadLock
+     */
+    @Trivial
+    void acquireReadLock() {
+        rwKeyStoreLock.readLock().lock();
+    }
+
+    /**
+     * Release the read lock. To be used to prevent concurrent keystore
+     * fetching and storing. Must be used with acquireReadLock
+     */
+    @Trivial
+    void releaseReadLock() {
+        rwKeyStoreLock.readLock().unlock();
+    }
+
+    /**
+     * Clone the keystore. If an exception occurs, return the original
+     * keystore as "best effort" instead of failing.
+     *
+     * @param original
+     * @return
+     */
+    @Trivial
+    private KeyStore cloneKeystore(KeyStore original) {
+        try {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            original.store(baos, password.getChars());
+            KeyStore clone = JSSEProviderFactory.getInstance().getKeyStoreInstance(type, provider);
+            clone.load(new ByteArrayInputStream(baos.toByteArray()), password.getChars());
+            return clone;
+        } catch (Throwable t) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(tc, "cloneKeystore hit an exception, will return original keystore.", t);
+            }
+        }
+        return original;
+    }
 }

--- a/dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/internal/KeyStoreServiceImpl.java
+++ b/dev/com.ibm.ws.ssl/src/com/ibm/ws/ssl/internal/KeyStoreServiceImpl.java
@@ -345,10 +345,7 @@ public class KeyStoreServiceImpl implements KeyStoreService {
                 throw new KeyStoreException("The keystore [" + keyStoreName + "] is not present in the configuration");
             } else {
                 wks.setCertificateEntry(alias, certificate);
-                wks.store();
             }
-        } catch (CertificateException e) {
-            throw e;
         } catch (KeyStoreException e) {
             throw e;
         } catch (Exception e) {
@@ -365,10 +362,7 @@ public class KeyStoreServiceImpl implements KeyStoreService {
                 throw new KeyStoreException("The keystore [" + keyStoreName + "] is not present in the configuration");
             } else {
                 wks.setKeyEntry(alias, key, chain);
-                wks.store();
             }
-        } catch (CertificateException e) {
-            throw e;
         } catch (KeyStoreException e) {
             throw e;
         } catch (Exception e) {

